jaxrs/WFLY-13530_expose_JAXRS_to_gRPC.adoc

```diff
@@ -0,0 +1,133 @@
+= Expose JAX-RS Resources to gRPC clients
+:author:            Ron Sigal
+:email:             rsigal@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+In the five years since gRPC was introduced, it has become a popular invocation
+and streaming framework, used by major corporations. Among its benefits are
+
+* speed
+* language independence
+
+With this feature request we propose to expose JAX-RS services as gRPC services,
+accessible by gRPC clients.
+
+Given a gRPC service definition like
+----
+service Greeter {
+  rpc SayHello (HelloRequest) returns (HelloReply) {}
+}
+----
+
+the protobuf java compiler will create a class GreeterGrpc, and a service can be
+created by subclassing. We envision making it as easy as something like this
+to forward a gRPC request to JAX-RS:
+
+----
+
+static class GreeterImpl extends GreeterGrpc.GreeterImplBase {
+
+   public void sayHello(HelloRequest req, StreamObserver<HelloReply> responseObserver) {
+      JAXRSForwarderBuilder builder = new JAXRSForwarderBuilder();
+      builder.servlet("ResteasyServlet").pathTranslator((String s) -> ("test/" + s));
+      JAXRSForwarder forwarder = builder.build();
+      forwarder.forward(req, (StreamObserver) responseObserver);
+   }
+}
+----
+
+The JAXRSForwarder will find the appropriate servlet, create the appropriate HTTP
+requests and responses, call the JAX-RS resource method, and send
+the result to the gRPC StreamObserver. 
+
+== Issue Metadata
+
+=== Issue
+
+* {issue-base-url}/EAP7-1536 [EAP7-1536]
+* {issue-base-url}/WFLY-13530 [WFLY-13530]
+
+
+=== Related Issues
+
+* {issue-base-url}/PRODMGT-1811 [PRODMGT-1811]
+* {issue-base-url}/WFLY-11034 [WFLY-11034]
+* {issue-base-url}/RESTEASY-2656 [RESTEASY-2656]
+
+=== Dev Contacts
+
+* mailto:rsigal@redhat.com[Ron Sigal]
+* mailto:asoldano@redhat.com[Alessio Soldano]
+* mailto:ema@redhat.com[Jim Ma]
+* mailto:rsearls@redhat.com[Rebecca Searls]
+* mailto:weli@redhat.com[Wei Nan Li]
+* mailto:jperkins@redhat.com[James Perkins]
+
+=== QE Contacts
+
+* ???
+
+=== Testing By
+
+TBD
+
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* RESTEasy
+* Wildfly
+
+=== Other Interested Projects
+
+* Elytron ?
+
+== Requirements
+
+=== Hard Requirements
+
+* A gRPC server will be exposed on its own port.
```

Comment:
gRPC is also new to me so I am trying to perform some research for the security aspects but it seems later on HTTP/2 - wouldn't that mean it can come over the existing Undertow connectors?

Reply:
Hi @darranl,

Well, that's a good question. The Java gRPC framework runs on Netty, so I supposed that we would need to hook it into Netty. When I communicated with Flavia, I posed the issue in those terms. I can go back and see what she says about using the current Undertow.

-Ron

Reply:
@ronsigal if some discussions have already happened lets try and get ongoing discussions into maybe a more open thread somewhere - I think gRPC has a lot of implications for the server as a whole.

Reply:
Ok, sure. Until recently, I was just trying to figure out what was even possible. In fact, mostly I was just looking at protobuf as a transport layer until Bilge mention gRPC at the F2F meeting. So the previouse discussion has mostly been me spewing a bunch of stuff at the RESTEasy folks (sorry, RESTEasy folks). But now that it's exposed to the light of day, it's probably time to be more open. How would you suggest proceeding?

Reply:
Just to link this from another conversation @jimma linked https://github.com/grpc/grpc-java/pull/4738. If that can become part of the grpc-java project that would allow us to use that instead.

Reply:
Thanks, @jamezp, and thank you @jimma for investigating that project.

It seems that they're doing the inverse of what I've been working on. That is, they're catching a gRPC request, dispatching it to a servlet, constructing a gRPC request, and dispatching it to a gRPC server. It sounds at first like there's some redundancy there, but probably not. What I'm doing is subclassing a gRPC class to catch a gRPC request, constructing an HttpServletRequest and an HttpServletResponse, and dispatching it to a JAX-RS resource by way of a RESTEasy servlet.

It seems that, given one of their servlets, we could dispatch either to a gRPC service or a JAX-RS "service" (resource). 

This is going to take some investigation. 

I've just pushed some experimental code to https://github.com/ronsigal/grpc-java. In particular, there are three classes in https://github.com/ronsigal/grpc-java/tree/grpc-jaxrs/examples/src/main/java/io/grpc/examples/helloworld:
1. HelloWorldServerJAXRS is a modification of the gRPC server HelloWorldServer.java.
2. JAXRSForwarderBuilder builds an instance of
3. JAXRSForwarder, which constructs an getHttpServletRequest and an getHttpServletRequest and passes them to RESTEasy.

It's very preliminary.